File: functions_waterClassification_Zou2018.py
import ee

def Ndvi(image):
    # // calculate ndvi
    ndvi = image.normalizedDifference(['Nir', 'Red']).rename('ndvi')
    return ndvi

def Evi(image):
    # calculate the enhanced vegetation index
    evi = image.expression('2.5 * (Nir - Red) / (1 + Nir + 6 * Red - 7.5 * Blue)', {
    'Nir': image.select(['Nir']),
    'Red': image.select(['Red']),
    'Blue': image.select(['Blue'])
    })
    return evi.rename(['evi'])

def Mndwi(image):
    # calculate mndwi
    mndwi = image.normalizedDifference(['Green', 'Swir1']).rename('mndwi')
    return mndwi

def ClassifyWaterZou2018(image):
    mndwi = Mndwi(image)
    ndvi = Ndvi(image)
    evi = Evi(image)

    water = (mndwi.gt(ndvi).Or(mndwi.gt(evi))).And(evi.lt(0.1))
    return(water.rename(['waterMask']))

File: RivWidthCloud_GEE_REPO/functions_Landsat578/functions_waterClassification_Zou2018.js
// Calculate water mask according to Zou et al., 2018
// https://www.pnas.org/content/pnas/suppl/2018/03/20/1719275115.DCSupplemental/pnas.1719275115.sapp.pdf

var Ndvi = function(image) {
  // calculate normalized difference vegetation index
  var ndvi = image.normalizedDifference(['Nir', 'Red']).rename('ndvi');
  return(ndvi);
};
var Evi = function(image) {
  // calculate the enhanced vegetation index
  var evi = image.expression('2.5 * (Nir - Red) / (1 + Nir + 6 * Red - 7.5 * Blue)', {
    'Nir': image.select(['Nir']),
    'Red': image.select(['Red']),
    'Blue': image.select(['Blue'])
    });
  return(evi.rename(['evi']));
};
var Mndwi = function(image) {
  // calculate modified normalized difference water index
  var mndwi = image.normalizedDifference(['Green', 'Swir1']).rename('mndwi');
  return(mndwi);
};

exports.ClassifyWaterZou2018 = function(image) {
  
  var mndwi = Mndwi(image);
  var ndvi = Ndvi(image);
  var evi = Evi(image);
  var water = (mndwi.gt(ndvi).or(mndwi.gt(evi))).and(evi.lt(0.1)); 
  
  return(water.rename(['waterMask']));
};